Forms: Move Salesforce settings to block modal

edanzer opened the pull request:
## Proposed changes:

Adds a Salesforce settings card to the Forms block integrations modal. Specific changes:
- Adds the SalesforceCard component to the forms modal
- Hides the Salesforce panel from the inspector sidebar 
- Updates the integrations endpoint to return Salesforce data.
   - NOTE: We do not track any Salesforce setting globally for the site, so this essentially returning hard coded data. This is something we should change in the future. You should be able to add a Salesforce organization ID globally, and have it be available for all forms. But that's not the way the sales force integration works now. 
- Adjusts IntegrationCardHeader toggle to show for all integrations even when plugin/integration is not installed or connected. In these cases, it shows but is disabled. 

<img width="1247" alt="salesforce-card" src="https://github.com/user-attachments/assets/1a62bf70-1b78-453e-998c-47f727afc4ba" />

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
None

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Test normal forms
   - Insert a regular form and open the integrations modal
   - Confirm the salesforce card shows
   - The toggle on the header should be disabled initially. It only enables when you add a valid Saleforce ID is added
   - Add any 15-18 digit string to the Organization ID field. Confirm the toggle on the header is enabled. 
   - Turn on the toggle on the header. Confirm the salesforce active icon shows on the sidebar panel. 
   - Try adding a non-valid Org ID (less than 15), click out of the input, and confirm you get a validation error. 
   - There's explanatory text in the body of the card that will disappear once an ID is added. 
* Test Salesforce form variant
   - Insert a Saleforce form
   - Confirm the Salesforce panel no longer shows on the panel
   - Everything should otherwise work the same as above, except as soon as you add your org ID, the header toggle will be enabled.

edanzer committed:
Add salesforce to forms modal

edanzer committed:
changelog

edanzer commented on projects/packages/forms/src/blocks/contact-form/components/jetpack-integrations-modal/active-integrations/index.js:
```diff
@@ -28,6 +29,19 @@ export default function ActiveIntegrations( { integrations, attributes, isLoadin
+						tooltip: __( 'Salesforce is connected for this form', 'jetpack-forms' ),
+					} );
+				}
+				break;
```
If Salesforce ID is valid and sendToSalesforce is true in block settings, we'll show the active salesforce icon in the Integrations panel.

edanzer commented on projects/packages/forms/src/blocks/contact-form/components/jetpack-integrations-modal/integration-card/integration-card-body.js:
```diff
@@ -5,18 +5,26 @@ const IntegrationCardBody = ( { isExpanded, children, cardData = {} } ) => {
+
+	const showPluginInstallMessage = isPlugin && ! isInstalled;
+	const showPluginActivateMessage = isPlugin && isInstalled && ! isActive;
+	const showContent = ( isPlugin && isInstalled && isActive ) || isService;
```
I had to adjust these conditionals to handle services, not just plugins. As the conditionals were looking complex, I just pulled them out and defined them up here to use in the return statement.

edanzer commented on projects/packages/forms/src/blocks/contact-form/components/jetpack-integrations-modal/integration-card/integration-card-header.js:
```diff
@@ -96,13 +96,13 @@ const IntegrationCardHeader = ( {
+									checked={ headerToggleValue && ( isActive || isConnected ) }
 									onChange={ handleToggleChange }
-									disabled={ ! isHeaderToggleEnabled }
+									disabled={ ! isHeaderToggleEnabled || ! ( isActive || isConnected ) }
```
This is a generic improvement that was needed for salesforce but improves the card header. Before: if plugin was not active or service not connected, we would not show a toggle to enable the integration for this form. Now: we will show the toggle, but it will be disabled/off.

edanzer commented on projects/packages/forms/src/blocks/contact-form/components/jetpack-integrations-modal/salesforce-card.js:
```diff
@@ -0,0 +1,93 @@
+	const [ organizationIdError, setOrganizationIdError ] = useState( false );
+	const isValidOrgId = !! (
+		salesforceData.organizationId &&
+		salesforceData.organizationId.trim().match( /^[a-zA-Z0-9]{15,18}$/ )
```
This is how the original code for salesforce validates the organization ID. There is a bit of duplication here, then on line 42, and active-integrations/index.js on line 36. I should centralize the logic for validating the ID. Note that we don't ever check this is a real id at salesforce - it's a simple validation check.

edanzer replied:
Add a single method for this.

edanzer commented on projects/packages/forms/src/blocks/contact-form/edit.js:
```diff
@@ -222,14 +215,6 @@ function JetpackContactFormEdit( { name, attributes, setAttributes, clientId, cl
-							setAttributes={ setAttributes }
-							instanceId={ instanceId }
-						/>
-					) }
```
Since Salesforce settings are managed via the integrations modal, we no longer need this sidebar panel. I have not yet deleted the actual component code for this. I was going to wait until we release this and have it out for a bit first.

edanzer commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -603,22 +608,56 @@ public function get_all_integrations_status( $request ) {
+					'details'     => array(),
+				);
+		}
+	}
```
I set up a generic get_service_status() method to handle non-plugin services like Google, Zapier, Salesforce, etc. The logic to get their status and return them would live here. At present, we do not keep track of Salesforce connection status at the site-level, only via the block editor. That's something we should change,  but it's a different PR.

edanzer committed:
Add single method to test salesforce id

edanzer committed:
Add explanatory salesforce text

edanzer committed:
Adjust card markup

simison commented:
I don't know if these logos are better but couple SVGs that I could find from Calypso:

https://github.com/Automattic/wp-calypso/blob/8747a6725bec900c837af8447d0cfde27e758ff8/packages/plans-grid-next/src/components/features-grid/client-logo-list/client-list.tsx#L61C1-L64C10

https://github.com/Automattic/wp-calypso/blob/8747a6725bec900c837af8447d0cfde27e758ff8/client/assets/images/logos/salesforce.svg?plain=1

simison commented on projects/packages/forms/src/blocks/contact-form/components/jetpack-integrations-modal/salesforce-card.js:
```diff
@@ -0,0 +1,99 @@
+
+	return (
+		<IntegrationCard
+			title={ __( 'Salesforce', 'jetpack-forms' ) }
```
No need to translate brand name.

edanzer replied:
Updated.

edanzer replied:
Actually, removing this generated an automated testing error on the PR around translatable strings. So restoring the translation. It's superfluous but also does not harm.

simison replied:
Let's check with @Automattic/i18n if they have advice. Was it a linter error that an ignore line could solve?

Not a blocker for merging.

edanzer replied:
I didn't screenshot the error in the testing log, but yes, I think it was linter error flagging non-translated strings.

dlind1 replied:
I'm not familiar with Jetpack linters, but in this case its fine to disable it for that specific line.

simison replied:
Untranslated again to see the error (my local linter doesn't show it), and then I'll add lint-ignore.

simison replied:
Hmm:
```
[packages/forms 1:03.112] ERROR in editor.js: Optimization seems to have broken the following translation strings:
[packages/forms 1:03.112]  - "Salesforce"
[packages/forms 1:03.112] 
[packages/forms 1:03.112] webpack 5.94.0 compiled with 1 error and 2 warnings in 35893 ms
[packages/forms 1:03.250]  ELIFECYCLE  Command failed with exit code 1.
[packages/forms 1:03.363]  ELIFECYCLE  Command failed with exit code 1.
[packages/forms 1:03.449]  ELIFECYCLE  Command failed with exit code 1.
[packages/forms 1:03.545] Script pnpm run build-production handling the build-production event returned with error code 1
[packages/forms 1:03.585] 
[packages/forms 1:03.586] Build failed: Error: Command failed with exit code 1: composer run --timeout=0 build-production
[packages/forms 1:03.586]     at makeError (file:///home/runner/work/jetpack/jetpack/node_modules/.pnpm/execa@7.0.0/node_modules/execa/lib/error.js:59:11)
[packages/forms 1:03.587]     at handlePromise (file:///home/runner/work/jetpack/jetpack/node_modules/.pnpm/execa@7.0.0/node_modules/execa/index.js:119:26)
[packages/forms 1:03.588]     at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
[packages/forms 1:03.588]     at async buildProject (file:///home/runner/work/jetpack/jetpack/tools/cli/commands/build.js:778:3)
[packages/forms 1:03.591]     at async file:///home/runner/work/jetpack/jetpack/tools/cli/commands/build.js:360:7
```

simison replied:
Okay, I'll include translation here and remove it in follow-up :-)

edanzer committed:
Update Salesforce svg

edanzer commented:
> I don't know if these logos are better but couple SVGs that I could find from Calypso:

I did update the Salesforce SVG to the same pattern as in Calypso. Main difference is it says 'Salesforce' within the cloud. I tweaked to to be blue and take in size props.

edanzer committed:
Restore Salesforce translation due to test error

simison commented:
Seeing console error:

<img width="942" alt="Screenshot 2025-05-07 at 17 21 58" src="https://github.com/user-attachments/assets/d92785b7-980a-4a0f-ad18-31361b21218f" />

Could the input invalidate only onBlur when there's actual content? Empty inputs shouldn't invalidate.
<img width="699" alt="Screenshot 2025-05-07 at 17 10 29" src="https://github.com/user-attachments/assets/e77c7864-7edd-453e-bf00-86543fddb6cd" />

The widow in the end looks off and spacing between link and input is too small;
<img width="676" alt="Screenshot 2025-05-07 at 17 23 35" src="https://github.com/user-attachments/assets/6263573d-c28c-4089-bfb7-a29f233e2789" />

I was thinking that widow we can either fix by applying this CSS to whole modal, which can be done in separate PR and possibly use a mixin from Jetpack's base-styles package:
```h1,
h2,
h3,
h4,
h5,
h6 {
	clear: both;
	text-wrap: balance; // TODO: remove this fallback once `pretty` is globally supported.
	text-wrap: pretty;
}

caption,
figcaption {
	text-wrap: balance;
	text-wrap: pretty;
}

p,
ul,
ol,
blockquote {
	text-wrap: pretty;
}
```

...or by using `Text` component:
```js
import {
	__experimentalText as Text, // eslint-disable-line @wordpress/no-unsafe-wp-apis
} from '@wordpress/components';
```

The help text could be in text input directly via `help={}` prop? Or just have its own `<p>` to ensure good spacing.

simison commented:
Another thought: the disabled toggle could have a tooltip explaining why it's disabled, i.e. that they need to add the ID to be able to enable it.

edanzer committed:
Update based on PR feedback

edanzer committed:
Fix salesforce svg

edanzer commented:
@simison and others, I've made the following changes based on feedback: 
- Avoid showing validationn error when org id field is empty
- Updated nudge in card body (given we now have tooltip too), and also added css to avoid orphaned words
- Added tooltip so users know what they need to do to enable the toggle on the header
   - This involve small adjustment to card header component too
- Added spacing between input and link to Salesforce information
- Updated Salesforce svg to remove the black on the letters

Now looks like:
<img width="676" alt="salesforce-card-2" src="https://github.com/user-attachments/assets/9bfcccf0-b1ab-4e25-a093-3208049bd1b9" />

edanzer committed:
Ensure salesforce id is always a string

simison committed:
Untranslate Salesforce and use en-dash for number range

simison reviewed (dismissed):
Tests well! Thanks for the last round of fixes. Tested setting ID, tested desktop and mobile versions.

> Seeing console error:

Confirmed that's in trunk so no need fixing here.

I'm just [pushing](https://github.com/Automattic/jetpack/pull/43297/commits/3bb0ac7de167384edbefe8572e65276cb336b045) two copy changes; en-dash for number range and ~untranslating "Salesforce"~ and after those this is good to go.

simison committed:
Put back Salesforce translation
